### PR TITLE
NIAD-587: Condition missing elements 

### PR DESCRIPTION
**Pull request by Broom94**

### CODE REVIEW

**Review by bartek-sarul — DISMISSED** [review]

File: service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/mapper/ConditionMapper.java [hunk 1]
```diff
@@ -13,28 +15,43 @@
...
+        }
+
         condition
             .setClinicalStatus(Condition.ConditionClinicalStatus.ACTIVE)
             .setVerificationStatus(Condition.ConditionVerificationStatus.UNKNOWN)
             .setSubject(encounter.getSubject())
             .setContext(new Reference(encounter))
-            .setAsserter(encounter.getParticipantFirstRep().getIndividual())
-            .setAssertedDate(assertedDate);
+            .setAssertedDate(assertedDate)
```
Comment (dawidskainos) [inline service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/mapper/ConditionMapper.java:51]: We should populate full CodeableConcept - system, code and display
See for detailed values: http://hl7.org/fhir/stu3/valueset-languages.html

**Review by dawidskainos — CHANGES_REQUESTED** [review]

File: service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/enums/Language.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+    private final String system = "http://hl7.org/fhir/ValueSet/languages";
+    private final String code;
+    private final String display;
+
+    public static uk.nhs.adaptors.oneoneone.cda.report.enums.Language fromCode(String code) {
+        return Stream.of(values())
+            .filter(am -> code.toLowerCase().equals(am.code))
+            .findFirst()
+            .orElse(null);
```
Comment (bartek-sarul) [inline service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/enums/Language.java:75]: [optional] A good modern practice is not to return null but Optional or just throw exception that value does not exist

File: service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/enums/MaritalStatus.java [hunk 3]
```diff
@@ -20,13 +20,13 @@
...
+    private final String code;
     private final String display;
 
     public static uk.nhs.adaptors.oneoneone.cda.report.enums.MaritalStatus fromCode(String code) {
         return Stream.of(values())
-            .filter(am -> code.toUpperCase().equals(am.value))
+            .filter(am -> code.toUpperCase().equals(am.code))
             .findFirst()
-            .orElseThrow();
+            .orElse(null);
```
Comment (bartek-sarul) [inline service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/enums/MaritalStatus.java:30]: [optional] A good modern practice is not to return null but Optional or just throw exception that value does not exist

File: service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/mapper/ConditionMapper.java [hunk 4]
```diff
@@ -5,35 +5,56 @@
...
 
+        for (POCDMT000002UK01Component5 component : section.getComponentArray()) {
+            if (component.getSection() != null) {
+                if (component.getSection().isSetLanguageCode()) {
+                    if (component.getSection().getLanguageCode().isSetCode()) {
+                        if (Language.fromCode(component.getSection().getLanguageCode().getCode()) != null) {
+                            Language language = Language.fromCode(component.getSection().getLanguageCode().getCode());
+                            condition.setCode(new CodeableConcept(
+                                new Coding()
+                                    .setDisplay(language.getDisplay())
```
Comment (bartek-sarul) [inline service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/mapper/ConditionMapper.java:43]: Each line below can throw NPE as `fromCode` can return null

File: service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/mapper/PatientMapper.java [hunk 5]
```diff
@@ -71,7 +73,19 @@ public Patient mapPatient(POCDMT000002UK01PatientRole patientRole) {
...
 
             if (itkPatient.isSetMaritalStatusCode()) {
-                fhirPatient.setMaritalStatus(getMaritalStatus(itkPatient));
+                if (itkPatient.getMaritalStatusCode().isSetCode()) {
+                    if (MaritalStatus.fromCode(itkPatient.getMaritalStatusCode().getCode()) != null) {
+                        MaritalStatus maritalStatus = MaritalStatus.fromCode(itkPatient.getMaritalStatusCode().getCode());
+                        if (maritalStatus.getDisplay() != null) {
+                            fhirPatient.setMaritalStatus(new CodeableConcept(
+                                new Coding()
+                                    .setDisplay(maritalStatus.getDisplay())
```
Comment (bartek-sarul) [inline service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/mapper/PatientMapper.java:82]: Each line below can throw NPE as fromCode can return null

**Review by bartek-sarul — CHANGES_REQUESTED** [review]

**Review by dawidskainos — APPROVED** [review]

File: service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/mapper/ConditionMapper.java [hunk 6]
```diff
@@ -3,37 +3,67 @@
...
         Date assertedDate = DateUtil.parse(itkEncounter.getEffectiveTime().getValue());
 
+        for (POCDMT000002UK01Component5 component : section.getComponentArray()) {
+            if (component.getSection() != null) {
+                if (component.getSection().isSetLanguageCode()) {
+                    if (component.getSection().getLanguageCode().isSetCode()) {
+                        Optional<Language> language = Language.fromCode(component.getSection().getLanguageCode().getCode());
+                        language.ifPresent(lang -> {
+                            Coding coding = new Coding();
+                            if (!StringUtils.isBlank(language.get().getDisplay())) {
```
Comment (bartek-sarul) [inline service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/mapper/ConditionMapper.java:44]: Don't use `language.get()`. You already have the optional unwrapped in `lang` variable.

File: service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/mapper/ConditionMapper.java [hunk 7]
```diff
@@ -3,37 +3,67 @@
...
 
         condition.setIdElement(newRandomUuid());
 
         Date assertedDate = DateUtil.parse(itkEncounter.getEffectiveTime().getValue());
 
+        for (POCDMT000002UK01Component5 component : section.getComponentArray()) {
+            if (component.getSection() != null) {
+                if (component.getSection().isSetLanguageCode()) {
+                    if (component.getSection().getLanguageCode().isSetCode()) {
+                        Optional<Language> language = Language.fromCode(component.getSection().getLanguageCode().getCode());
```
Comment (bartek-sarul) [inline service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/mapper/ConditionMapper.java:41]: You can inline this into 
```
Language.fromCode(component.getSection().getLanguageCode().getCode())
    .ifPresent(language -> { ...
```

File: service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/mapper/PatientMapper.java [hunk 8]
```diff
@@ -71,7 +75,26 @@ public Patient mapPatient(POCDMT000002UK01PatientRole patientRole) {
             }
 
             if (itkPatient.isSetMaritalStatusCode()) {
-                fhirPatient.setMaritalStatus(getMaritalStatus(itkPatient));
+                if (itkPatient.getMaritalStatusCode().isSetCode()) {
+                    if (MaritalStatus.fromCode(itkPatient.getMaritalStatusCode().getCode()) != null) {
+                        Optional<MaritalStatus> maritalStatus = MaritalStatus.fromCode(itkPatient.getMaritalStatusCode().getCode());
```
Comment (bartek-sarul) [inline service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/mapper/PatientMapper.java:80]: The same comment as in `ConditionMapper`

File: service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/mapper/PatientMapper.java [hunk 9]
```diff
@@ -71,7 +75,26 @@ public Patient mapPatient(POCDMT000002UK01PatientRole patientRole) {
...
+                        maritalStatus.ifPresent(mart -> {
+                            Coding coding = new Coding();
+                            if (!StringUtils.isBlank(maritalStatus.get().getCode())) {
```
Comment (bartek-sarul) [inline service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/mapper/PatientMapper.java:83]: The same comment as in `ConditionMapper`

**Review by bartek-sarul — CHANGES_REQUESTED** [review]

**Review by bartek-sarul — APPROVED** [review]

**Review by dawidskainos — APPROVED** [review]

**Pull request merged**